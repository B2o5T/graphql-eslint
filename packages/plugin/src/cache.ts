// Based on the `eslint-plugin-import`'s cache
// https://github.com/import-js/eslint-plugin-import/blob/main/utils/ModuleCache.js
import debugFactory from 'debug';

const log = debugFactory('graphql-eslint:ModuleCache');

export class ModuleCache<K, T> {
  map = new Map<K, { lastSeen: [number, number]; result: T }>();

  set(cacheKey: K, result: T): void {
    this.map.set(cacheKey, { lastSeen: process.hrtime(), result });
    log('setting entry for', cacheKey);
  }

  get(cacheKey: K, settings = { lifetime: 10 /* seconds */ }): T | void {
    const value = this.map.get(cacheKey);
    if (!value) {
      log('cache miss for', cacheKey);
      return;
    }
    const { lastSeen, result } = value;
    // check freshness
    if (
      process.env.NODE /* don't check for ESLint CLI */ ||
      process.hrtime(lastSeen)[0] < settings.lifetime
    ) {
      return result;
    }
  }
}
