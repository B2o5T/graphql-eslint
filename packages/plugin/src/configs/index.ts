import { ConfigName } from '../types.js';
// @ts-expect-error -- complains about no default export
import operationsAllConfig from './operations-all.js';
// @ts-expect-error -- complains about no default export
import operationsRecommendedConfig from './operations-recommended.js';
// @ts-expect-error -- complains about no default export
import schemaAllConfig from './schema-all.js';
// @ts-expect-error -- complains about no default export
import schemaRecommendedConfig from './schema-recommended.js';
// @ts-expect-error -- complains about no default export
import relayConfig from './schema-relay.js';

export const configs = {
  'schema-recommended': schemaRecommendedConfig,
  'schema-all': schemaAllConfig,
  'schema-relay': relayConfig,
  'operations-recommended': operationsRecommendedConfig,
  'operations-all': operationsAllConfig,
  'flat/schema-recommended': schemaRecommendedConfig.rules,
  'flat/schema-all': {
    ...schemaRecommendedConfig.rules,
    ...schemaAllConfig.rules,
  },
  'flat/schema-relay': relayConfig.rules,
  'flat/operations-recommended': operationsRecommendedConfig.rules,
  'flat/operations-all': {
    ...operationsRecommendedConfig.rules,
    ...operationsAllConfig.rules,
  },
} satisfies Record<ConfigName | `flat/${ConfigName}`, unknown>;
