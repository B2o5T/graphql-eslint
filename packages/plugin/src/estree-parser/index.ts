export * from "./converter";
export * from "./estree-ast";
export * from "./utils";
