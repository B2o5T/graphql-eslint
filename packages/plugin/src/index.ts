import { configs } from './configs/index.js';
import { parseForESLint, parser } from './parser.js';
import { processor } from './processor.js';
import { rules } from './rules/index.js';

export * from './types.js';
export type { IGraphQLConfig } from 'graphql-config';
export type { GraphQLTagPluckOptions } from '@graphql-tools/graphql-tag-pluck';

export { requireGraphQLSchemaFromContext, requireSiblingsOperations } from './utils.js';

export const processors = { graphql: processor };

export { rules, configs, parser, parseForESLint };

// eslint-disable-next-line import/no-default-export -- It's common practice for ESLint plugins that supports Flat config to use the default export
export default {
  parser,
  processor,
  rules,
  configs,
};
