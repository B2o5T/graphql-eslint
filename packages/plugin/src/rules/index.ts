/*
 * ðŸš¨ IMPORTANT! Do not manually modify this file. Run: `pnpm generate:configs`
 */

import { rule as alphabetize } from './alphabetize/index.js';
import { rule as descriptionStyle } from './description-style/index.js';
import { GRAPHQL_JS_VALIDATIONS } from './graphql-js-validation.js';
import { rule as inputName } from './input-name/index.js';
import { rule as loneExecutableDefinition } from './lone-executable-definition/index.js';
import { rule as matchDocumentFilename } from './match-document-filename/index.js';
import { rule as namingConvention } from './naming-convention/index.js';
import { rule as noAnonymousOperations } from './no-anonymous-operations/index.js';
import { rule as noDeprecated } from './no-deprecated/index.js';
import { rule as noDuplicateFields } from './no-duplicate-fields/index.js';
import { rule as noHashtagDescription } from './no-hashtag-description/index.js';
import { rule as noOnePlaceFragments } from './no-one-place-fragments/index.js';
import { rule as noRootType } from './no-root-type/index.js';
import { rule as noScalarResultTypeOnMutation } from './no-scalar-result-type-on-mutation/index.js';
import { rule as noTypenamePrefix } from './no-typename-prefix/index.js';
import { rule as noUnreachableTypes } from './no-unreachable-types/index.js';
import { rule as noUnusedFields } from './no-unused-fields/index.js';
import { rule as relayArguments } from './relay-arguments/index.js';
import { rule as relayConnectionTypes } from './relay-connection-types/index.js';
import { rule as relayEdgeTypes } from './relay-edge-types/index.js';
import { rule as relayPageInfo } from './relay-page-info/index.js';
import { rule as requireDeprecationDate } from './require-deprecation-date/index.js';
import { rule as requireDeprecationReason } from './require-deprecation-reason/index.js';
import { rule as requireDescription } from './require-description/index.js';
import { rule as requireFieldOfTypeQueryInMutationResult } from './require-field-of-type-query-in-mutation-result/index.js';
import { rule as requireImportFragment } from './require-import-fragment/index.js';
import { rule as requireNullableFieldsWithOneof } from './require-nullable-fields-with-oneof/index.js';
import { rule as requireNullableResultInRoot } from './require-nullable-result-in-root/index.js';
import { rule as requireSelections } from './require-selections/index.js';
import { rule as requireTypePatternWithOneof } from './require-type-pattern-with-oneof/index.js';
import { rule as selectionSetDepth } from './selection-set-depth/index.js';
import { rule as strictIdInTypes } from './strict-id-in-types/index.js';
import { rule as uniqueEnumValueNames } from './unique-enum-value-names/index.js';
import { rule as uniqueFragmentName } from './unique-fragment-name/index.js';
import { rule as uniqueOperationName } from './unique-operation-name/index.js';

export const rules = {
  ...GRAPHQL_JS_VALIDATIONS,
  alphabetize,
  'description-style': descriptionStyle,
  'input-name': inputName,
  'lone-executable-definition': loneExecutableDefinition,
  'match-document-filename': matchDocumentFilename,
  'naming-convention': namingConvention,
  'no-anonymous-operations': noAnonymousOperations,
  'no-deprecated': noDeprecated,
  'no-duplicate-fields': noDuplicateFields,
  'no-hashtag-description': noHashtagDescription,
  'no-one-place-fragments': noOnePlaceFragments,
  'no-root-type': noRootType,
  'no-scalar-result-type-on-mutation': noScalarResultTypeOnMutation,
  'no-typename-prefix': noTypenamePrefix,
  'no-unreachable-types': noUnreachableTypes,
  'no-unused-fields': noUnusedFields,
  'relay-arguments': relayArguments,
  'relay-connection-types': relayConnectionTypes,
  'relay-edge-types': relayEdgeTypes,
  'relay-page-info': relayPageInfo,
  'require-deprecation-date': requireDeprecationDate,
  'require-deprecation-reason': requireDeprecationReason,
  'require-description': requireDescription,
  'require-field-of-type-query-in-mutation-result': requireFieldOfTypeQueryInMutationResult,
  'require-import-fragment': requireImportFragment,
  'require-nullable-fields-with-oneof': requireNullableFieldsWithOneof,
  'require-nullable-result-in-root': requireNullableResultInRoot,
  'require-selections': requireSelections,
  'require-type-pattern-with-oneof': requireTypePatternWithOneof,
  'selection-set-depth': selectionSetDepth,
  'strict-id-in-types': strictIdInTypes,
  'unique-enum-value-names': uniqueEnumValueNames,
  'unique-fragment-name': uniqueFragmentName,
  'unique-operation-name': uniqueOperationName,
};
