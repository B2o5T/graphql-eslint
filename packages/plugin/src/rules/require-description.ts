import { ASTKindToNode, Kind, TokenKind } from 'graphql';
import type { GraphQLESLintRule, ValueOf } from '../types';
import { getLocation, TYPES_KINDS } from '../utils';
import type { GraphQLESTreeNode } from '../estree-converter';

const RULE_ID = 'require-description';

const ALLOWED_KINDS = [
  ...TYPES_KINDS,
  Kind.DIRECTIVE_DEFINITION,
  Kind.FIELD_DEFINITION,
  Kind.INPUT_VALUE_DEFINITION,
  Kind.ENUM_VALUE_DEFINITION,
  Kind.OPERATION_DEFINITION,
] as const;

type AllowedKind = typeof ALLOWED_KINDS[number];
type AllowedKindToNode = Pick<ASTKindToNode, AllowedKind>;
type SelectorNode = GraphQLESTreeNode<ValueOf<AllowedKindToNode>>;

export type RequireDescriptionRuleConfig = {
  types?: boolean;
  operationFieldDefinition?: boolean;
} & {
    [key in AllowedKind]?: boolean;
  };

function getNodeName(node: SelectorNode) {
  const DisplayNodeNameMap = {
    [Kind.OBJECT_TYPE_DEFINITION]: 'type',
    [Kind.INTERFACE_TYPE_DEFINITION]: 'interface',
    [Kind.ENUM_TYPE_DEFINITION]: 'enum',
    [Kind.SCALAR_TYPE_DEFINITION]: 'scalar',
    [Kind.INPUT_OBJECT_TYPE_DEFINITION]: 'input',
    [Kind.UNION_TYPE_DEFINITION]: 'union',
    [Kind.DIRECTIVE_DEFINITION]: 'directive',
  } as const;

  switch (node.kind) {
    case Kind.OBJECT_TYPE_DEFINITION:
    case Kind.INTERFACE_TYPE_DEFINITION:
    case Kind.ENUM_TYPE_DEFINITION:
    case Kind.SCALAR_TYPE_DEFINITION:
    case Kind.INPUT_OBJECT_TYPE_DEFINITION:
    case Kind.UNION_TYPE_DEFINITION:
      return `${DisplayNodeNameMap[node.kind]} ${node.name.value}`;
    case Kind.DIRECTIVE_DEFINITION:
      return `${DisplayNodeNameMap[node.kind]} @${node.name.value}`;
    case Kind.FIELD_DEFINITION:
    case Kind.INPUT_VALUE_DEFINITION:
    case Kind.ENUM_VALUE_DEFINITION:
      return `${node.parent.name.value}.${node.name.value}`;
    case Kind.OPERATION_DEFINITION:
      return node.name ? `${node.operation} ${node.name.value}` : node.operation;
  }
}

const rule: GraphQLESLintRule<[RequireDescriptionRuleConfig]> = {
  meta: {
    docs: {
      category: 'Schema',
      description: 'Enforce descriptions in type definitions and operations.',
      url: `https://github.com/B2o5T/graphql-eslint/blob/master/docs/rules/${RULE_ID}.md`,
      examples: [
        {
          title: 'Incorrect',
          usage: [{ types: true, FieldDefinition: true }],
          code: /* GraphQL */ `
            type someTypeName {
              name: String
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ types: true, FieldDefinition: true }],
          code: /* GraphQL */ `
            """
            Some type description
            """
            type someTypeName {
              """
              Name description
              """
              name: String
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ OperationDefinition: true }],
          code: /* GraphQL */ `
            # Create a new user
            mutation createUser {
              # ...
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ operationFieldDefinition: true }],
          code: /* GraphQL */ `
            type Mutation {
              "Create a new user"
              createUser: User
            }
          `,
        },
      ],
      configOptions: [
        {
          types: true,
          [Kind.DIRECTIVE_DEFINITION]: true,
        },
      ],
      recommended: true,
    },
    type: 'suggestion',
    messages: {
      [RULE_ID]: 'Description is required for `{{ nodeName }}`.',
    },
    schema: {
      type: 'array',
      minItems: 1,
      maxItems: 1,
      items: {
        type: 'object',
        additionalProperties: false,
        minProperties: 1,
        properties: {
          types: {
            type: 'boolean',
            description: `Includes:\n\n${TYPES_KINDS.map(kind => `- \`${kind}\``).join('\n')}`,
          },
          operationFieldDefinition: {
            type: 'boolean',
            description: 'Definitions within Query, Mutation, and Subscription'
          },
          ...Object.fromEntries(
            [...ALLOWED_KINDS].sort().map(kind => {
              let description = `Read more about this kind on [spec.graphql.org](https://spec.graphql.org/October2021/#${kind}).`;
              if (kind === Kind.OPERATION_DEFINITION) {
                description +=
                  '\n\n> You must use only comment syntax `#` and not description syntax `"""` or `"`.';
              }
              return [kind, { type: 'boolean', description }];
            }),
          ),
        },
      },
    },
  },
  create(context) {
    const { types, operationFieldDefinition, ...restOptions } = context.options[0] || {};

    const kinds = new Set<string>(types ? TYPES_KINDS : []);
    for (const [kind, isEnabled] of Object.entries(restOptions)) {
      if (isEnabled) {
        kinds.add(kind);
      } else {
        kinds.delete(kind);
      }
    }

    if (operationFieldDefinition) {
      kinds.add(':matches(ObjectTypeDefinition, ObjectTypeExtension)[name.value=/Query|Mutation|Subscription/] > FieldDefinition');
    }

    const selector = [...kinds].join(',');

    return {
      [selector](node: SelectorNode) {
        let description = '';
        const isOperation = node.kind === Kind.OPERATION_DEFINITION;
        if (isOperation) {
          const rawNode = node.rawNode();
          const { prev, line } = rawNode.loc.startToken;
          if (prev.kind === TokenKind.COMMENT) {
            const value = prev.value.trim();
            const linesBefore = line - prev.line;
            if (!value.startsWith('eslint') && linesBefore === 1) {
              description = value;
            }
          }
        } else {
          description = node.description?.value.trim() || '';
        }

        if (description.length === 0) {
          context.report({
            loc: isOperation ? getLocation(node.loc.start, node.operation) : node.name.loc,
            messageId: RULE_ID,
            data: {
              nodeName: getNodeName(node),
            },
          });
        }
      },
    };
  },
};

export default rule;
