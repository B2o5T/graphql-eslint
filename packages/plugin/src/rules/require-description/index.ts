import { ASTKindToNode, Kind, TokenKind } from 'graphql';
import { getRootTypeNames } from '@graphql-tools/utils';
import { GraphQLESTreeNode } from '../../estree-converter/index.js';
import { GraphQLESLintRule, ValueOf } from '../../types.js';
import {
  ARRAY_DEFAULT_OPTIONS,
  eslintSelectorsTip,
  getLocation,
  getNodeName,
  requireGraphQLSchema,
  TYPES_KINDS,
} from '../../utils.js';

export const RULE_ID = 'require-description';

const ALLOWED_KINDS = [
  ...TYPES_KINDS,
  Kind.DIRECTIVE_DEFINITION,
  Kind.FIELD_DEFINITION,
  Kind.INPUT_VALUE_DEFINITION,
  Kind.ENUM_VALUE_DEFINITION,
  Kind.OPERATION_DEFINITION,
] as const;

type AllowedKind = (typeof ALLOWED_KINDS)[number];
type AllowedKindToNode = Pick<ASTKindToNode, AllowedKind>;
type SelectorNode = GraphQLESTreeNode<ValueOf<AllowedKindToNode>>;

const schema = {
  type: 'array',
  minItems: 1,
  maxItems: 1,
  items: {
    type: 'object',
    additionalProperties: false,
    minProperties: 1,
    properties: {
      types: {
        type: 'boolean',
        enum: [true],
        description: `Includes:\n${TYPES_KINDS.map(kind => `- \`${kind}\``).join('\n')}`,
      },
      rootField: {
        type: 'boolean',
        enum: [true],
        description: 'Definitions within `Query`, `Mutation`, and `Subscription` root types.',
      },
      ignoredSelectors: {
        ...ARRAY_DEFAULT_OPTIONS,
        description: ['Ignore specific selectors', eslintSelectorsTip].join('\n'),
      },
      ...Object.fromEntries(
        [...ALLOWED_KINDS].sort().map(kind => {
          let description = `> [!NOTE]
>
> Read more about this kind on [spec.graphql.org](https://spec.graphql.org/October2021/#${kind}).`;
          if (kind === Kind.OPERATION_DEFINITION) {
            description += [
              '',
              '',
              '> [!WARNING]',
              '>',
              '> You must use only comment syntax `#` and not description syntax `"""` or `"`.',
            ].join('\n');
          }
          return [kind, { type: 'boolean', description }];
        }),
      ),
    },
  },
} as const;

// TODO try import { FromSchema } from 'json-schema-to-ts';
export type RuleOptions = [
  {
    [key in AllowedKind]?: boolean;
  } & {
    types?: true;
    rootField?: true;
    ignoredSelectors?: string[];
  },
];

export const rule: GraphQLESLintRule<RuleOptions> = {
  meta: {
    docs: {
      category: 'Schema',
      description: 'Enforce descriptions in type definitions and operations.',
      url: `https://the-guild.dev/graphql/eslint/rules/${RULE_ID}`,
      examples: [
        {
          title: 'Incorrect',
          usage: [{ types: true, FieldDefinition: true }],
          code: /* GraphQL */ `
            type someTypeName {
              name: String
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ types: true, FieldDefinition: true }],
          code: /* GraphQL */ `
            """
            Some type description
            """
            type someTypeName {
              """
              Name description
              """
              name: String
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ OperationDefinition: true }],
          code: /* GraphQL */ `
            # Create a new user
            mutation createUser {
              # ...
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ rootField: true }],
          code: /* GraphQL */ `
            type Mutation {
              "Create a new user"
              createUser: User
            }

            type User {
              name: String
            }
          `,
        },
        {
          title: 'Correct',
          usage: [
            {
              ignoredSelectors: [
                '[type=ObjectTypeDefinition][name.value=PageInfo]',
                '[type=ObjectTypeDefinition][name.value=/(Connection|Edge)$/]',
              ],
            },
          ],
          code: /* GraphQL */ `
            type FriendConnection {
              edges: [FriendEdge]
              pageInfo: PageInfo!
            }
            type FriendEdge {
              cursor: String!
              node: Friend!
            }
            type PageInfo {
              hasPreviousPage: Boolean!
              hasNextPage: Boolean!
              startCursor: String
              endCursor: String
            }
          `,
        },
      ],
      configOptions: [
        {
          types: true,
          [Kind.DIRECTIVE_DEFINITION]: true,
          rootField: true,
        },
      ],
      recommended: true,
    },
    type: 'suggestion',
    messages: {
      [RULE_ID]: 'Description is required for {{ nodeName }}',
    },
    schema,
  },
  create(context) {
    let { types, rootField, ignoredSelectors = [], ...restOptions } = context.options[0] || {};

    const kinds = new Set<string>(types ? TYPES_KINDS : []);
    for (const [kind, isEnabled] of Object.entries(restOptions)) {
      if (isEnabled) {
        kinds.add(kind);
      } else {
        kinds.delete(kind);
      }
    }

    if (rootField) {
      const schema = requireGraphQLSchema(RULE_ID, context);
      const rootTypeNames = getRootTypeNames(schema);
      kinds.add(
        `:matches(ObjectTypeDefinition, ObjectTypeExtension)[name.value=/^(${[
          ...rootTypeNames,
        ].join(',')})$/] > FieldDefinition`,
      );
    }
    let selector = `:matches(${[...kinds]})`;
    for (const str of ignoredSelectors) {
      selector += `:not(${str})`;
    }
    return {
      [selector](node: SelectorNode) {
        let description = '';
        const isOperation = node.kind === Kind.OPERATION_DEFINITION;
        if (isOperation) {
          const rawNode = node.rawNode();
          const { prev, line } = rawNode.loc!.startToken;
          if (prev?.kind === TokenKind.COMMENT) {
            const value = prev.value.trim();
            const linesBefore = line - prev.line;
            if (!value.startsWith('eslint') && linesBefore === 1) {
              description = value;
            }
          }
        } else {
          description = node.description?.value.trim() || '';
        }

        if (description.length === 0) {
          context.report({
            loc: isOperation ? getLocation(node.loc.start, node.operation) : node.name.loc,
            messageId: RULE_ID,
            data: {
              nodeName: getNodeName(node),
            },
          });
        }
      },
    };
  },
};
