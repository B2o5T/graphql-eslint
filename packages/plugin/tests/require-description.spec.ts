import { ParserOptions } from '../src';
import { RuleTester } from '@theguild/eslint-rule-tester';
import { rule, RuleOptions, RULE_ID } from '../src/rules/require-description';

const ruleTester = new RuleTester();
const OPERATION = { OperationDefinition: true };

ruleTester.run<RuleOptions>('require-description', rule, {
  valid: [
    {
      code: /* GraphQL */ `
        enum EnumUserLanguagesSkill {
          """
          basic
          """
          basic
          """
          fluent
          """
          fluent
          """
          native
          """
          native
        }
      `,
      options: [{ EnumValueDefinition: true }],
    },
    {
      code: /* GraphQL */ `
        input SalaryDecimalOperatorsFilterUpdateOneUserInput {
          """
          gt
          """
          gt: BSONDecimal
          """
          in
          """
          in: [BSONDecimal]
          " nin "
          nin: [BSONDecimal]
        }
      `,
      options: [{ InputValueDefinition: true }],
    },
    {
      code: /* GraphQL */ `
        " Test "
        type CreateOneUserPayload {
          "Created document ID"
          recordId: MongoID

          "Created document"
          record: User
        }
      `,
      options: [{ types: true, FieldDefinition: true }],
    },
    {
      code: /* GraphQL */ `
        # OK
        query {
          test
        }
      `,
      options: [OPERATION],
    },
    {
      code: /* GraphQL */ `
        # ignore fragments
        fragment UserFields on User {
          id
        }
      `,
      options: [OPERATION],
    },
    {
      ...useSchema(/* GraphQL */ `
        type Query {
          "Get user"
          user(id: ID!): User
        }
      `),
      options: [{ rootField: true }],
    },
    {
      ...useSchema('type Query'),
      options: [{ rootField: true }],
    },
  ],
  invalid: [
    {
      code: 'type User { id: ID }',
      options: [{ ObjectTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'interface Node { id: ID! }',
      options: [{ InterfaceTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'enum Role { ADMIN }',
      options: [{ EnumTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'scalar Email',
      options: [{ ScalarTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'input CreateUserInput { email: Email! }',
      options: [{ InputObjectTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'union Media = Book | Movie',
      options: [{ UnionTypeDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'directive @auth(requires: Role!) on FIELD_DEFINITION',
      options: [{ DirectiveDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'type User { email: Email! }',
      options: [{ FieldDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'input CreateUserInput { email: Email! }',
      options: [{ InputValueDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      code: 'enum Role { ADMIN }',
      options: [{ EnumValueDefinition: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      name: 'should disable description for ObjectTypeDefinition',
      code: /* GraphQL */ `
        type CreateOneUserPayload {
          recordId: MongoID
          record: User
        }
      `,
      options: [
        {
          types: true,
          ObjectTypeDefinition: false,
          FieldDefinition: true,
        },
      ],
      errors: [{ messageId: RULE_ID }, { messageId: RULE_ID }],
    },
    {
      name: 'should report because of linesBefore !== 1',
      code: /* GraphQL */ `
        # linesBefore !== 1

        query {
          foo
        }
      `,
      options: [OPERATION],
      errors: [{ messageId: RULE_ID }],
    },
    {
      name: 'should validate mutation',
      code: 'mutation createUser { foo }',
      options: [OPERATION],
      errors: [{ messageId: RULE_ID }],
    },
    {
      name: 'should validate subscription',
      code: 'subscription commentAdded { foo }',
      options: [OPERATION],
      errors: [{ messageId: RULE_ID }],
    },
    {
      name: 'should report because skips comment that starts with `eslint`',
      code: /* GraphQL */ `
        # eslint-disable-next-line semi
        query {
          foo
        }
      `,
      options: [OPERATION],
      errors: [{ messageId: RULE_ID }],
    },
    {
      name: 'should ignore comments before fragment definition',
      code: /* GraphQL */ `
        # BAD
        fragment UserFields on User {
          id
        }

        query {
          user {
            ...UserFields
          }
        }
      `,
      options: [OPERATION],
      errors: [{ messageId: RULE_ID }],
    },
    {
      ...useSchema('type Query { user(id: String!): User! }'),
      options: [{ rootField: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      ...useSchema('type Mutation { createUser(id: [ID!]!): User! }'),
      options: [{ rootField: true }],
      errors: [{ messageId: RULE_ID }],
    },
    {
      ...useSchema(/* GraphQL */ `
        type MySubscription {
          users: [User!]!
        }
        schema {
          subscription: MySubscription
        }
      `),
      options: [{ rootField: true }],
      errors: [{ messageId: RULE_ID }],
    },
  ],
});

function useSchema(code: string): { code: string; parserOptions: Pick<ParserOptions, 'schema'> } {
  return {
    code,
    parserOptions: {
      schema: /* GraphQL */ `
        type User {
          id: ID!
        }

        ${code}
      `,
    },
  };
}
