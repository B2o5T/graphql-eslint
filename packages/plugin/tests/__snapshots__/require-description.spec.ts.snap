// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
❌ Error 1/1

> 1 | type User { id: ID }
    |      ^^^^ Description is required for \`type User\`.
`;

exports[` 2`] = `
❌ Error 1/1

> 1 | interface Node { id: ID! }
    |           ^^^^ Description is required for \`interface Node\`.
`;

exports[` 3`] = `
❌ Error 1/1

> 1 | enum Role { ADMIN }
    |      ^^^^ Description is required for \`enum Role\`.
`;

exports[` 4`] = `
❌ Error 1/1

> 1 | scalar Email
    |        ^^^^^ Description is required for \`scalar Email\`.
`;

exports[` 5`] = `
❌ Error 1/1

> 1 | input CreateUserInput { email: Email! }
    |       ^^^^^^^^^^^^^^^ Description is required for \`input CreateUserInput\`.
`;

exports[` 6`] = `
❌ Error 1/1

> 1 | union Media = Book | Movie
    |       ^^^^^ Description is required for \`union Media\`.
`;

exports[` 7`] = `
❌ Error 1/1

> 1 | directive @auth(requires: Role!) on FIELD_DEFINITION
    |            ^^^^ Description is required for \`directive @auth\`.
`;

exports[` 8`] = `
❌ Error 1/1

> 1 | type User { email: Email! }
    |             ^^^^^ Description is required for \`User.email\`.
`;

exports[` 9`] = `
❌ Error 1/1

> 1 | input CreateUserInput { email: Email! }
    |                         ^^^^^ Description is required for \`CreateUserInput.email\`.
`;

exports[` 10`] = `
❌ Error 1/1

> 1 | enum Role { ADMIN }
    |             ^^^^^ Description is required for \`Role.ADMIN\`.
`;

exports[` 11`] = `
❌ Error 1/2

  1 |
  2 |         type CreateOneUserPayload {
> 3 |           recordId: MongoID
    |           ^^^^^^^^ Description is required for \`CreateOneUserPayload.recordId\`.
  4 |           record: User
  5 |         }
  6 |       

❌ Error 2/2

  1 |
  2 |         type CreateOneUserPayload {
  3 |           recordId: MongoID
> 4 |           record: User
    |           ^^^^^^ Description is required for \`CreateOneUserPayload.record\`.
  5 |         }
  6 |       
`;

exports[` 12`] = `
❌ Error 1/1

  1 |
  2 |         # linesBefore !== 1
  3 |
> 4 |         query {
    |         ^^^^^ Description is required for \`query\`.
  5 |           foo
  6 |         }
  7 |       
`;

exports[` 13`] = `
❌ Error 1/1

> 1 | mutation createUser { foo }
    | ^^^^^^^^ Description is required for \`mutation createUser\`.
`;

exports[` 14`] = `
❌ Error 1/1

> 1 | subscription commentAdded { foo }
    | ^^^^^^^^^^^^ Description is required for \`subscription commentAdded\`.
`;

exports[` 15`] = `
❌ Error 1/1

  1 |
  2 |         # eslint-disable-next-line semi
> 3 |         query {
    |         ^^^^^ Description is required for \`query\`.
  4 |           foo
  5 |         }
  6 |       
`;

exports[` 16`] = `
❌ Error 1/1

   1 |
   2 |         # BAD
   3 |         fragment UserFields on User {
   4 |           id
   5 |         }
   6 |
>  7 |         query {
     |         ^^^^^ Description is required for \`query\`.
   8 |           user {
   9 |             ...UserFields
  10 |           }
  11 |         }
  12 |       
`;
